### DISCUSSION TITLE
Question: How can I implement/emulate a java package class (for example java.lang.System)

### DISCUSSION THREAD

**Discussion by McMurat** [xpenatan/gdx-teavm#132] [General]
Hi,
I really like this project, good work! I want to migrate my existing libgdx application and need to implement some additional classlib classes/methods which are missing in TeaVM. How can I do this? I have seen that there is an @Emulate annotation in gdx-teavm and there is also a separate directory of emulated classes. How can I do this for my own project?

**Comment by xpenatan** [comment]
The emulate anotation is a solution I got it to work that replace the original code with a custom. the magic is inside TeaClassTransformer class plugin.  What it does it remove everything from the original class and replace it with the class that use the Emulate annotation during the build process.  All methods must contain the same identity (name, return and params) as the original class.

There is also the teavm way of doing this. It is  by using resources/META-INF/teavm.properties file.  You can see here https://github.com/xpenatan/gdx-imgui/blob/master/imgui/imgui-teavm/src/main/resources/META-INF/teavm.properties. There are some examples of this in teavm test folders.

`mapPackageHierarchy|gen.imgui=imgui` 
What this do is when teavm code is parsed. Everything inside gen.input package will replace classes in imgui package.

You wont see any imgui class in github because its all generated.

**Reply by xpenatan** [reply]
You want to add missing methods that teavm dont have or implement the entire class?

**Reply by McMurat** [reply]
Thank you for your answer and the hint towards teavm.properties.
I guess I will need both. Implement some additional missing methods and entire missing classes.

**Reply by xpenatan** [reply]
Emulate annotation can help by replacing just few parts of the original clase code by using `updateCode` to true.  Its a bit limited and I only used in a few classes like ClassEmu and FloatBufferOverArrayEmu for example. For this you need to add Emulate annotation in everything.

**Reply by McMurat** [reply]
But to use this annotation I will still have to setup the teavm.properties file or how does the compiler know about the emulated files?

**Reply by xpenatan** [reply]
No, The @emulate is a custom solution. teavm.properties is teavm solution.

teavm-backend is using Reflection library to search the emulate annotation in these packages ("emu", "com", "org", "net").

https://github.com/xpenatan/gdx-teavm/blob/master/backends/backend-teavm/src/main/java/com/github/xpenatan/gdx/backends/teavm/config/plugins/TeaClassTransformer.java#L73-L81

**Comment by McMurat** [comment]
Thank you, got it working!
